Hoxhunt integration v2

tomaskukk opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Description
New version of Hoxhunt integration. Changed support from community from to partner.t

Includes
- Deprecation of old integration
- New integration
- Default playbook
- Incident types
- Incident fields
- Classifier, mapper incoming & outgoing
- Default layout
- Mirroring

## Must have
- [x] Tests
- [x] Documentation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=37019) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=37019) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=37019) it.</sub>

yaakovpraisler reviewed (changes requested):
Hey @tomaskukk Thank you for your contribution and sorry for the delay with the review.
Overall it looks good, please see my comments below.

@idovandijk You can start reviewing the non-integration content.

tomaskukk commented on Packs/Hoxhunt/Integrations/HoxhuntV2/HoxhuntV2.yml:
```diff
@@ -0,0 +1,217 @@
+  type: 0
+  required: false
+  additionalinfo: (<number> <time unit>, e.g., 12 hours, 7 days). Defaults to 7 days.
+  defaultvalue: "7 days"
```
The amount of incidents varies a lot, I think this is a good default, if a customer would have huge amounts of incidents per day, we'll still [apply the max fetch limit](https://github.com/demisto/content/pull/37019/files/d1cf6e062845c8a28ee0eae571fc8f973d70b94b#diff-c58c6b5e7529cb2f06b87629bf7883a0aadaa86799f0679a22f7c558c82b3d28R490) so it's safe even if there are a lot of threats for n days

tomaskukk commented on Packs/Hoxhunt/Integrations/HoxhuntV2/HoxhuntV2.yml:
```diff
@@ -0,0 +1,217 @@
+      description: The ID of the Hoxhunt incident the threats will be from.
+    - name: query_options
+      description: >
+        Optional string for query options. Inserted directly into the query with following validation in place: first (max 100, default 50), sort (default createdAt_DESC), and filter (default {}).  Example: `first: 10, filter: {state_eq: OPEN}, sort: createdAt_DESC`  Refer to the Hoxhunt API documentation for more information.
```
Good point, I agree with the idea of providing simple ways to use the command. We'll refactor!

tomaskukk commented on Packs/Hoxhunt/Integrations/HoxhuntV2/HoxhuntV2.yml:
```diff
@@ -0,0 +1,217 @@
+      type: string
+    description: Updates Incident state.
+  - description: Get mapping fields from remote incident. Please note that this method will not update the current incident. It's here for debugging purposes.
+    name: get-mapping-fields
```
I understood that it's optional to add those based on the [docs](https://xsoar.pan.dev/docs/integrations/mirroring_integration#debugging). I didn't find the need to include them. We can add them if necessary, otherwise I'd just skip having them exposed.

yaakovpraisler replied:
Ok, I was just wondering because the `get-mapping-fields` is there.

tomaskukk commented on Packs/Hoxhunt/Integrations/HoxhuntV2/HoxhuntV2.py:
```diff
@@ -0,0 +1,926 @@
+    }
+
+    state = new_incident_data.get('state')
+    if state in entry_actions:
```
Not necessarily, but it doesn't seem to affect the incident in any way if the incident is open in Cortex XSOAR. So having it doesn't seem to cause any harm.

idovandijk commented:
Thank you @tomaskukk for your patience, I will send my review in the upcoming days.

yaakovpraisler commented on Packs/Hoxhunt/Integrations/HoxhuntV2/HoxhuntV2.py:
```diff
@@ -0,0 +1,858 @@
+
+
+class Client(BaseClient):
+    def __init__(self, base_url: str, headers: dict, proxy: bool = False, verify: bool = False):
+        super().__init__(base_url=base_url, headers=headers, proxy=proxy, verify=verify)
```
```suggestion
```
When __init__ is not defined, the super.__init__ will be run

yaakovpraisler commented on Packs/Hoxhunt/ReleaseNotes/2_0_0.md:
```diff
@@ -0,0 +1,217 @@
+##### New: Hoxhunt-classifier
+
+-   New: Classifies Hoxhunt incidents
+    <~XSOAR> (Available from Cortex XSOAR 6.10.0).</~XSOAR>
```
```suggestion
```
Sorry, I missed this in my previous comment here, I would do it by my self but I don't have permissions.
Please do it for all the RN entries.

yaakovpraisler reviewed (commented):
Hey @tomaskukk, Let's wait for @idovandijk's review for the incident types, fields etc. then we'll set a call for a demo.

idovandijk reviewed (dismissed):
Looks good! I have a few comments regarding the fields:
1. There are many new fields in this PR. While some fields are Hoxhunt specific, there are some fields that could potentially be replaced with existing fields in the marketplace.

Please take a look at the existing incident fields in the CommonTypes pack in the Marketplace.
If you find a suitable replacement field, you can associate it to your Hoxhunt incident types by adding your incident types under **both** the `associatedTypes` **and** under the `systemAssociatedTypes` properties of the field in its JSON file.

If the new field you want to use is a content pack other than CommonTypes, you can decide to keep your existing field, or to use the existing one and add a dependency in your pack to the destination pack. Since using fields from packs other than CommonTypes (which is preinstalled) will introduce a new dependency for your pack, I leave this to you to decide. CommonTypes however, is a Core pack, which means XSOAR users get those fields preinstalled anyway, so it's worth using them.

The advantage to using existing fields rather than adding new ones is that they are shared across different vendors and can be used in widgets, dashboards, etc. For example, if I want to filter all incidents by the "Email To" fields, I won't be seeing incidents where the "Hoxhunt Email To" field is populated.

2. In addition to my previous point regarding fields, there is an additional check I'd like to ask from your side. As the Hoxhunt pack is also available in XSIAM, we can use Field Aliases instead of adding those new alert fields into XSIAM (we will only add them to XSOAR). Please go over the CoreAlertFields pack, and see whether any fields from there can be utilized for your use-case.

For example, let's say the Email Recipient field in the CoreAlertFields pack fulfills our need for our Hox FT Email To field. We will do the following:
A. Add the name of the field Hox FT Email To under the `Aliases` property in the field Email Recipient from the CoreAlertFields pack like so:
![image](https://github.com/user-attachments/assets/9404ff15-8e29-4c26-9c6e-a698ab56f38a)
B. Edit the Hox FT Email To field and add the following property, to make sure this field is added only to XSOAR (because XSIAM will use the alias);
```
"marketplaces": [
"xsoar"
]
```

3. Please rename the fields to have spaces between words in Title Case.

Regarding the playbook: Before using `rasterize` add a task with the `IsIntegrationAvailable` automation to check if Rasterize is available prior to using it.

Once you're able to introduce the changes, I would be happy to see the playbook in action, the mapper and classifier, and the layout :)

If you have any questions please don't hesitate to let me know. Looking forward to get this merged 😎

tomaskukk commented on Packs/Hoxhunt/IncidentFields/incidentfield-HoxMessageToUsers.json:
```diff
@@ -0,0 +1,34 @@
+        "Hoxhunt User Acted"
+    ],
+    "associatedToAll": false,
+    "unmapped": false,
```
No, I can add `unsearchable: true`, good catch!

idovandijk commented:
> We also had this chat with @edik24, the reason we wanted to use custom fields is to keep the fields `isReadOnly: true`. These fields should never be edited by the user, as they are kept up to date within our server. User should have no control over them. The suitable replacements that I checked, were `isReadOnly: false`.
> 
> There are a couple of fields that are `isReadOnly: false`, but we didn't find suitable built-in system fields for them.

Thank you for bringing this to my attention. In that case, it makes sense to keep your fields.

> Honestly I'm not that familiar with XSIAM. I trust your guidance, but does it have implications such as mentioned above in the first comment reply (readonly fields)?

Let me get back to you this. As I don't want you to have to make changes and then revert them, please ignore this point for now until I can update with how it will behave.

One more thing, as you make changes to the field names, be aware that these changes will require also updating the mapper and layout to reflect the new names.

tomaskukk commented:
> Let me get back to you this. As I don't want you to have to make changes and then revert them, please ignore this point for now until I can update with how it will behave.

Alright thank you!

> One more thing, as you make changes to the field names, be aware that these changes will require also updating the mapper and layout to reflect the new names.

Yes! We renamed the field names already and updated the incoming mapper. Outgoing mapper and layouts are using the `cliName` which stayed the same with the changes.

Thank you for all the help so far! Let us know if there's any other changes we need to make out before scheduling the demo :)

tomaskukk commented:
Hey @idovandijk just wanted to ask if there is something that is expected from us to move this forwards now :)

idovandijk commented:
> Hey @idovandijk just wanted to ask if there is something that is expected from us to move this forwards now :)

Hi @tomaskukk, it looks like the behavior for field aliases is: the main field's `readOnly` property is propagated to the aliases configured in it. So in your case since your field is set to `readOnly` and the main field in XSIAM is editable, you can also skip that part of my comment :)

We can continue to demo

tomaskukk commented:
Awesome @idovandijk, would you be able to propose some times that work for you, for example three slots that we could choose from?

idovandijk commented:
> Awesome @idovandijk, would you be able to propose some times that work for you, for example three slots that we could choose from?

We have 

> Awesome @idovandijk, would you be able to propose some times that work for you, for example three slots that we could choose from?

We've reached out with an email, let us know if that works and if not we can find a new time. Thanks again.

tomaskukk commented:
I submitted two small commits that:

1. Fixes pack readme not to show contributors twice
2. Fixed a few incident fields that still had wrong name format

pennane commented:
Reordered the config settings as per previuous demo discussion, also generated the readme and merged from base to resolve conflict.

@yaakovpraisler

yaakovpraisler commented on Packs/Hoxhunt/Integrations/HoxhuntV2/HoxhuntV2.py:
```diff
@@ -0,0 +1,855 @@
+
+
+def create_output(results: dict[str, str], endpoint: str, key_field: str = '') -> CommandResults:
+    human_readable = tableToMarkdown('Hoxhunt results', results)
```
Please use the `headers` and the `headerTransform` function arguments in order to became the headers more readable.
https://xsoar.pan.dev/docs/reference/api/common-server-python#tabletomarkdown

yaakovpraisler commented on Packs/Hoxhunt/Integrations/HoxhuntV2/HoxhuntV2.yml:
```diff
@@ -0,0 +1,223 @@
+  type: 8
+  required: false
+  additionalinfo: "If enabled, only open incidents will be fetched."
+- display: Use system proxy settings
+  name: proxy
+  type: 8
+  required: false
+- display: Incident Mirroring Direction
+  name: mirror_direction
+  defaultvalue: Incoming And Outgoing
+  type: 15
+  required: false
+  section: Collect
+  additionalinfo: "Choose the direction to mirror the incident: Incoming (from Hoxhunt to Cortex XSOAR), Outgoing (from Cortex XSOAR to Hoxhunt), or Incoming and Outgoing (from/to Cortex XSOAR and Hoxhunt)."
+  options:
+  - None
+  - Incoming
+  - Outgoing
+  - Incoming And Outgoing
+  hidden:
+  - marketplacev2
+- display: Trust any certificate (not secure)
+  name: insecure
+  type: 8
+  required: false
```
```suggestion
- display: Incident Mirroring Direction
  name: mirror_direction
  defaultvalue: Incoming And Outgoing
  type: 15
  required: false
  additionalinfo: "Choose the direction to mirror the incident: Incoming (from Hoxhunt to Cortex XSOAR), Outgoing (from Cortex XSOAR to Hoxhunt), or Incoming and Outgoing (from/to Cortex XSOAR and Hoxhunt)."
  options:
  - None
  - Incoming
  - Outgoing
  - Incoming And Outgoing
  hidden:
  - marketplacev2
- display: Trust any certificate (not secure)
  name: insecure
  type: 8
  required: false
- display: Use system proxy settings
  name: proxy
  type: 8
  required: false
```
Move the `proxy` to the bottom and removed the "section: Collect" from mirror_direction.
The "section: Collect" should be for related fetch parameters like "IncidentType", "incidentFetchInterval", "first_fetch" etc.

yaakovpraisler commented on Packs/Hoxhunt/ReleaseNotes/2_0_0.json:
```diff
@@ -0,0 +1,4 @@
+{
```
I don't know why I didn't noticed this before, but you can remove this file because it's not breaking anything, customer that already using the v1 integration will can continue to use it until they're updating to v2.

pennane commented:
- added explicit section tags into yml (Collect for fetch parameters, Connect for connecting)
- removed release notes __json__ file
- updated create_output headers to be in title case


_example of output:_
![CleanShot 2024-12-13 at 16 11 44@2x](https://github.com/user-attachments/assets/af9edd3a-a06b-4bdf-b89d-c2d0a511dc45)

yaakovpraisler commented on Packs/Hoxhunt/Integrations/HoxhuntV2/HoxhuntV2.yml:
```diff
@@ -0,0 +1,235 @@
+  defaultvalue: Incoming And Outgoing
+  type: 15
+  required: false
+  section: Collect
```
```suggestion
```
There are 2 time the field `section`

yaakovpraisler replied:
```suggestion
```
This parameter has 2 times the field `section`, please remove this as it should be in Connect

yaakovpraisler commented on Packs/Hoxhunt/Playbooks/Hox-Enrich-Incident_README.md:
```diff
@@ -0,0 +1,37 @@
+
+---
+
+![Hoxhunt - Enrich Incident](../doc_files/Hoxhunt_-_Enrich_Incident.png)
```
The image file doesn't exist under the doc_files dir and hasn't the same file name.
Please move out the image file to the doc_files dir and fix the name and make sure you can see the image in the playbook readme in XSOAR.

yaakovpraisler reviewed (commented):
I found the following issues during the build failure, please fix them so we can move forward and merge the PR.

Also you need to sign the CLAassistance as described [here](https://github.com/demisto/content/pull/37019#issuecomment-2451783892). (both of you @tomaskukk and @pennane)

Thanks.

tomaskukk commented:
> I found the following issues in the build pre-commit failure, please fix them so we can move forward and merge the PR.
Also you need to sign the CLAassistance as described https://github.com/demisto/content/pull/37019#issuecomment-2451783892. (both of you @tomaskukk and @pennane)
Thanks.

Done @yaakovpraisler!

yaakovpraisler reviewed (commented):
The build still fails, you can go to the pre-commit step to see the errors.
Anyway, we need to fix the following -

To fix this:
`Packs/Hoxhunt/Playbooks/Hox-Enrich-Incident.yml: [RM116] - No playbook image found, please add playbook image`
Please add the [.pack-ignore](https://xsoar.pan.dev/docs/packs/packs-format#pack-ignore) file and add the following section:
```
[file:Hox-Enrich-Incident.yml]
ignore=RM116
```

pennane commented:
Sorry about recurring pre-commit check fails. There are some problems running the pre-commit hook locally so sadly these were not caught before requesting reviews.

But another thing, is upgrading to py 3.12 necessary? We faced some mirroring problems prior when trying to upgrade from 3.10 and that is why it still has been at that version. We can of course check it futher.

yaakovpraisler commented:
@pennane I didn't notice before but we're not supporting yet 3.12 (this is currently in progress) but I'm not sure this should impact the mirroring.
Anyway, let's upgrade the docker tag to `3.11.10.116949`. (editing my previous comment as well)
